File: bindings/uniffi/tests/swift/Tests/ZenEngineTests/ZenEngineTests.swift
import Foundation
import Testing
@testable import ZenUniffi

enum TestDataError: Error {
    case notFound
}

func testDataRoot() throws -> URL {
    var current = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
    while current.path != "/" {
        let candidate = current.appendingPathComponent("test-data")
        if FileManager.default.fileExists(atPath: candidate.path) {
            return candidate
        }
        current = current.deletingLastPathComponent()
    }
    throw TestDataError.notFound
}

func readTestFile(_ name: String) throws -> Data {
    try Data(contentsOf: testDataRoot().appendingPathComponent(name))
}

func json(_ buffer: Data) throws -> [String: Any] {
    try #require(JSONSerialization.jsonObject(with: buffer) as? [String: Any])
}

final class FilesystemCallback: ZenDecisionLoaderCallback {
    let root: URL

    init(root: URL) {
        self.root = root
    }

    func load(key: String) async throws -> JsonBuffer? {
        try? Data(contentsOf: root.appendingPathComponent(key))
    }
}

final class SumNodeCallback: ZenCustomNodeCallback, @unchecked Sendable {
    var seenRequest: ZenEngineHandlerRequest?

    func handle(key: ZenEngineHandlerRequest) async throws -> ZenEngineHandlerResponse {
        seenRequest = key
        let input = try JSONSerialization.jsonObject(with: key.input) as? [String: Any]
        let a = input?["a"] as? Int ?? 0
        let output = try JSONSerialization.data(withJSONObject: ["data": a + 20])
        return ZenEngineHandlerResponse(output: output, traceData: nil)
    }
}

struct ZenEngineTests {
    @Test func staticLoader() async throws {
        let loader = ZenLoader.static(content: ["table.json": try readTestFile("table.json")])
        let engine = try ZenEngine(loader: loader)
        let response = try await engine.evaluate(
            key: "table.json",
            context: Data("{\"input\":12}".utf8),
            options: nil
        )
        #expect(try json(response.result)["output"] as? Int == 10)
        #expect(response.trace == nil)
    }

    @Test func filesystemLoader() async throws {
        let engine = try ZenEngine(loader: .filesystem(path: testDataRoot().path))
        let response = try await engine.evaluate(
            key: "table.json",
            context: Data("{\"input\":5}".utf8),
            options: nil
        )
        #expect(try json(response.result)["output"] as? Int == 0)
    }

    @Test func invalidZipFailsOnConstruction() {
        #expect(throws: (any Error).self) {
            try ZenEngine(loader: .zip(bytes: Data([1, 2, 3, 4])))
        }
    }

    @Test func callbackLoader() async throws {
        let callback = FilesystemCallback(root: try testDataRoot())
        let engine = try ZenEngine(loader: .callback(callback: callback))
        let response = try await engine.evaluate(
            key: "table.json",
            context: Data("{\"input\":12}".utf8),
            options: nil
        )
        #expect(try json(response.result)["output"] as? Int == 10)

        await #expect(throws: (any Error).self) {
            try await engine.evaluate(key: "missing.json", context: Data("{}".utf8), options: nil)
        }
    }

    @Test func missingKeyFails() async throws {
        let engine = try ZenEngine(loader: .static(content: [:]))
        await #expect(throws: (any Error).self) {
            try await engine.evaluate(key: "missing.json", context: Data("{}".utf8), options: nil)
        }
    }

    @Test func createDecision() async throws {
        let engine = try ZenEngine()
        let decision = try engine.createDecision(content: try readTestFile("table.json"))
        try decision.validate()
        let response = try await decision.evaluate(context: Data("{\"input\":12}".utf8), options: nil)
        #expect(try json(response.result)["output"] as? Int == 10)
    }

    @Test func getDecision() async throws {
        let engine = try ZenEngine(loader: .filesystem(path: testDataRoot().path))
        let decision = try await engine.getDecision(key: "table.json")
        let response = try await decision.evaluate(context: Data("{\"input\":12}".utf8), options: nil)
        #expect(try json(response.result)["output"] as? Int == 10)
    }

    @Test func evaluateBatch() async throws {
        let engine = try ZenEngine(loader: .filesystem(path: testDataRoot().path))
        let results = await engine.evaluateBatch(
            requests: [
                ZenBatchRequest(key: "table.json", context: Data("{\"input\":12}".utf8)),
                ZenBatchRequest(key: "missing.json", context: Data("{}".utf8)),
                ZenBatchRequest(key: "table.json", context: Data("{\"input\":5}".utf8)),
            ],
            options: nil
        )

        #expect(results.count == 3)
        #expect(results[0].success)
        #expect(try json(#require(results[0].data).result)["output"] as? Int == 10)
        #expect(!results[1].success)
        #expect(results[1].error != nil)
        #expect(results[2].success)
        #expect(try json(#require(results[2].data).result)["output"] as? Int == 0)
    }

    @Test func traceOption() async throws {
        let engine = try ZenEngine(loader: .filesystem(path: testDataRoot().path))
        let response = try await engine.evaluate(
            key: "table.json",
            context: Data("{\"input\":12}".utf8),
            options: ZenEvaluateOptions(maxDepth: 5, trace: true)
        )
        let trace = try #require(response.trace)
        #expect(!trace.isEmpty)
    }

    @Test func expressionGraph() async throws {
        let engine = try ZenEngine(loader: .filesystem(path: testDataRoot().path))
        let context = Data("{\"numbers\":[1,5,15,25],\"firstName\":\"John\",\"lastName\":\"Doe\"}".utf8)
        let response = try await engine.evaluate(key: "expression.json", context: context, options: nil)
        let result = try json(response.result)
        let expected = try json(Data(
            "{\"deep\":{\"nested\":{\"sum\":46}},\"fullName\":\"John Doe\",\"largeNumbers\":[15,25],\"smallNumbers\":[1,5]}".utf8
        ))
        #expect(result as NSDictionary == expected as NSDictionary)
    }

    @Test func functionGraph() async throws {
        let engine = try ZenEngine(loader: .filesystem(path: testDataRoot().path))
        let response = try await engine.evaluate(
            key: "function.json",
            context: Data("{\"input\":15}".utf8),
            options: nil
        )
        #expect(try json(response.result)["output"] as? Int == 30)
    }

    @Test func customNodeHandler() async throws {
        let handler = SumNodeCallback()
        let engine = try ZenEngine(loader: .filesystem(path: testDataRoot().path), customNode: handler)
        let response = try await engine.evaluate(
            key: "custom.json",
            context: Data("{\"a\":5}".utf8),
            options: nil
        )
        #expect(try json(response.result)["data"] as? Int == 25)

        let request = try #require(handler.seenRequest)
        #expect(request.node.kind == "sum")
        #expect(request.node.name == "customNode1")
        #expect(try json(request.node.config)["prop1"] as? String == "{{ a + 10 }}")
    }

    @Test func evaluateExpressionFunction() throws {
        let result = try evaluateExpression(
            expression: "sum(numbers)",
            context: Data("{\"numbers\":[1,2,3]}".utf8)
        )
        let value = try JSONSerialization.jsonObject(with: result, options: [.fragmentsAllowed])
        #expect(value as? Int == 6)
    }

    @Test func evaluateUnaryExpressionFunction() throws {
        #expect(try evaluateUnaryExpression(expression: "$ > 10", context: Data("{\"$\":15}".utf8)))
        #expect(try !evaluateUnaryExpression(expression: "$ > 10", context: Data("{\"$\":5}".utf8)))
    }

    @Test func invalidExpressionFails() {
        #expect(throws: (any Error).self) {
            try evaluateExpression(expression: "a +* b", context: Data("{}".utf8))
        }
    }

    @Test func compiledExpression() throws {
        let expression = try ZenExpression.compile(expression: "a + b")
        let result = try expression.evaluate(context: Data("{\"a\":1,\"b\":2}".utf8))
        let value = try JSONSerialization.jsonObject(with: result, options: [.fragmentsAllowed])
        #expect(value as? Int == 3)

        let unary = try ZenExpressionUnary.compile(expression: "$ > 3")
        #expect(try unary.evaluate(context: Data("{\"$\":4}".utf8)))
        #expect(try !unary.evaluate(context: Data("{\"$\":2}".utf8)))
    }
}

File: bindings/uniffi/tests/csharp/ZenEngineTests.cs
using System.IO.Compression;
using System.Text.Json.Nodes;
using GoRules.ZenEngine;
using Xunit;

namespace GoRules.ZenEngine.Tests;

public class ZenEngineTests
{
    private static string TestDataRoot()
    {
        var current = new DirectoryInfo(AppContext.BaseDirectory);
        while (current != null && !Directory.Exists(Path.Combine(current.FullName, "test-data")))
        {
            current = current.Parent;
        }

        if (current == null)
        {
            throw new InvalidOperationException("test-data directory not found");
        }

        return Path.Combine(current.FullName, "test-data");
    }

    private static JsonBuffer ReadTestFile(string name) =>
        new(File.ReadAllBytes(Path.Combine(TestDataRoot(), name)));

    private static JsonNode Json(JsonBuffer buffer) =>
        JsonNode.Parse(buffer.ToString())!;

    private class FilesystemCallback : ZenDecisionLoaderCallback
    {
        public Task<JsonBuffer?> Load(string key)
        {
            var path = Path.Combine(TestDataRoot(), key);
            return Task.FromResult<JsonBuffer?>(File.Exists(path) ? new JsonBuffer(File.ReadAllBytes(path)) : null);
        }
    }

    private class SumNodeCallback : ZenCustomNodeCallback
    {
        public ZenEngineHandlerRequest? SeenRequest;

        public Task<ZenEngineHandlerResponse> Handle(ZenEngineHandlerRequest key)
        {
            SeenRequest = key;
            var a = Json(key.Input)["a"]!.GetValue<int>();
            var output = new JsonBuffer($"{{\"data\":{a + 20}}}");
            return Task.FromResult(new ZenEngineHandlerResponse(output, null));
        }
    }

    [Fact]
    public async Task StaticLoader()
    {
        var loader = new ZenLoader.Static(new Dictionary<string, JsonBuffer>
        {
            ["table.json"] = ReadTestFile("table.json"),
        });

        using var engine = new ZenEngine(loader);
        var response = await engine.Evaluate("table.json", new JsonBuffer("{\"input\":12}"), null);
        Assert.Equal(10, Json(response.Result)["output"]!.GetValue<int>());
        Assert.Null(response.Trace);
    }

    [Fact]
    public async Task FilesystemLoader()
    {
        using var engine = new ZenEngine(new ZenLoader.Filesystem(TestDataRoot()));
        var response = await engine.Evaluate("table.json", new JsonBuffer("{\"input\":5}"), null);
        Assert.Equal(0, Json(response.Result)["output"]!.GetValue<int>());
    }

    [Fact]
    public async Task ZipLoader()
    {
        using var buffer = new MemoryStream();
        using (var zip = new ZipArchive(buffer, ZipArchiveMode.Create, true))
        {
            var entry = zip.CreateEntry("table.json");
            using var stream = entry.Open();
            stream.Write(ReadTestFile("table.json").Value);
        }

        using var engine = new ZenEngine(new ZenLoader.Zip(buffer.ToArray()));
        var response = await engine.Evaluate("table.json", new JsonBuffer("{\"input\":12}"), null);
        Assert.Equal(10, Json(response.Result)["output"]!.GetValue<int>());
    }

    [Fact]
    public void InvalidZipFailsOnConstruction()
    {
        Assert.ThrowsAny<ZenException>(() => new ZenEngine(new ZenLoader.Zip(new byte[] { 1, 2, 3, 4 })));
    }

    [Fact]
    public async Task CallbackLoader()
    {
        using var engine = new ZenEngine(new ZenLoader.Callback(new FilesystemCallback()));
        var response = await engine.Evaluate("table.json", new JsonBuffer("{\"input\":12}"), null);
        Assert.Equal(10, Json(response.Result)["output"]!.GetValue<int>());

        await Assert.ThrowsAnyAsync<ZenException>(() =>
            engine.Evaluate("missing.json", new JsonBuffer("{}"), null));
    }

    [Fact]
    public async Task MissingKeyFails()
    {
        using var engine = new ZenEngine(new ZenLoader.Static(new Dictionary<string, JsonBuffer>()));
        await Assert.ThrowsAnyAsync<ZenException>(() =>
            engine.Evaluate("missing.json", new JsonBuffer("{}"), null));
    }

    [Fact]
    public async Task CreateDecision()
    {
        using var engine = new ZenEngine();
        using var decision = engine.CreateDecision(ReadTestFile("table.json"));
        decision.Validate();
        var response = await decision.Evaluate(new JsonBuffer("{\"input\":12}"), null);
        Assert.Equal(10, Json(response.Result)["output"]!.GetValue<int>());
    }

    [Fact]
    public async Task GetDecision()
    {
        using var engine = new ZenEngine(new ZenLoader.Filesystem(TestDataRoot()));
        using var decision = await engine.GetDecision("table.json");
        var response = await decision.Evaluate(new JsonBuffer("{\"input\":12}"), null);
        Assert.Equal(10, Json(response.Result)["output"]!.GetValue<int>());
    }

    [Fact]
    public async Task EvaluateBatch()
    {
        using var engine = new ZenEngine(new ZenLoader.Filesystem(TestDataRoot()));
        var results = await engine.EvaluateBatch(new[]
        {
            new ZenBatchRequest("table.json", new JsonBuffer("{\"input\":12}")),
            new ZenBatchRequest("missing.json", new JsonBuffer("{}")),
            new ZenBatchRequest("table.json", new JsonBuffer("{\"input\":5}")),
        }, null);

        Assert.Equal(3, results.Length);
        Assert.True(results[0].Success);
        Assert.Equal(10, Json(results[0].Data!.Result)["output"]!.GetValue<int>());
        Assert.False(results[1].Success);
        Assert.NotNull(results[1].Error);
        Assert.True(results[2].Success);
        Assert.Equal(0, Json(results[2].Data!.Result)["output"]!.GetValue<int>());
    }

    [Fact]
    public async Task TraceOption()
    {
        using var engine = new ZenEngine(new ZenLoader.Filesystem(TestDataRoot()));
        var options = new ZenEvaluateOptions(5, true);
        var response = await engine.Evaluate("table.json", new JsonBuffer("{\"input\":12}"), options);
        Assert.NotNull(response.Trace);
        Assert.NotEmpty(response.Trace);
    }

    [Fact]
    public async Task ExpressionGraph()
    {
        using var engine = new ZenEngine(new ZenLoader.Filesystem(TestDataRoot()));
        var context = new JsonBuffer("{\"numbers\":[1,5,15,25],\"firstName\":\"John\",\"lastName\":\"Doe\"}");
        var response = await engine.Evaluate("expression.json", context, null);
        var expected = JsonNode.Parse(
            "{\"deep\":{\"nested\":{\"sum\":46}},\"fullName\":\"John Doe\",\"largeNumbers\":[15,25],\"smallNumbers\":[1,5]}");
        Assert.True(JsonNode.DeepEquals(expected, Json(response.Result)));
    }

    [Fact]
    public async Task FunctionGraph()
    {
        using var engine = new ZenEngine(new ZenLoader.Filesystem(TestDataRoot()));
        var response = await engine.Evaluate("function.json", new JsonBuffer("{\"input\":15}"), null);
        Assert.Equal(30, Json(response.Result)["output"]!.GetValue<int>());
    }

    [Fact]
    public async Task CustomNodeHandler()
    {
        var handler = new SumNodeCallback();
        using (var engine = new ZenEngine(new ZenLoader.Filesystem(TestDataRoot()), handler))
        {
            var response = await engine.Evaluate("custom.json", new JsonBuffer("{\"a\":5}"), null);
            Assert.Equal(25, Json(response.Result)["data"]!.GetValue<int>());
        }

        Assert.NotNull(handler.SeenRequest);
        Assert.Equal("sum", handler.SeenRequest!.Node.Kind);
        Assert.Equal("customNode1", handler.SeenRequest.Node.Name);
        Assert.Equal("{{ a + 10 }}", Json(handler.SeenRequest.Node.Config)["prop1"]!.GetValue<string>());
    }

    [Fact]
    public void EvaluateExpression()
    {
        var result = ZenUniffiMethods.EvaluateExpression("sum(numbers)", new JsonBuffer("{\"numbers\":[1,2,3]}"));
        Assert.Equal(6, Json(result).GetValue<int>());
    }

    [Fact]
    public void EvaluateUnaryExpression()
    {
        Assert.True(ZenUniffiMethods.EvaluateUnaryExpression("$ > 10", new JsonBuffer("{\"$\":15}")));
        Assert.False(ZenUniffiMethods.EvaluateUnaryExpression("$ > 10", new JsonBuffer("{\"$\":5}")));
    }

    [Fact]
    public void InvalidExpressionFails()
    {
        Assert.ThrowsAny<ZenException>(() =>
            ZenUniffiMethods.EvaluateExpression("a +* b", new JsonBuffer("{}")));
    }

    [Fact]
    public void CompiledExpression()
    {
        using (var expression = ZenExpression.Compile("a + b"))
        {
            var result = expression.Evaluate(new JsonBuffer("{\"a\":1,\"b\":2}"));
            Assert.Equal(3, Json(result).GetValue<int>());
        }

        using var unary = ZenExpressionUnary.Compile("$ > 3");
        Assert.True(unary.Evaluate(new JsonBuffer("{\"$\":4}")));
        Assert.False(unary.Evaluate(new JsonBuffer("{\"$\":2}")));
    }
}
